aten/src/ATen/core/type.cpp

```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+}
+
+bool UnionType::operator==(const Type& rhs) const {
+  if (auto union_rhs = rhs.cast<UnionType>()) {
+    return types_ == union_rhs->types_;
+  }
+  if (auto optional_rhs = rhs.cast<OptionalType>()) {
+    auto optional_lhs = this->to_optional();
+    if (optional_lhs) {
+      return *optional_rhs == *((optional_lhs.value())->cast<OptionalType>());
+    }
+    return false;
+  }
+  return false;
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (const auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors. We need to use
+    // `std::equal` with `TypeEqual` for optimal performance; do not
+    // change this to `operator==`!
+    if (std::equal(this->types_.begin(), this->types_.end(),
+                      union_rhs->types().begin(), union_rhs->types().end(),
+                      TypeEqual())) {
+      return true;
+    }
+    // Final O(n^2) check since the vector equality comparison above
+    // won't handle subtyping relationships
+    return std::all_of(this->types_.begin(), this->types_.end(),
+      [&](TypePtr this_type) -> bool {
+        return this_type->isSubtypeOf(union_rhs);
+    });
+  } else if (const auto optional_rhs = rhs_->cast<OptionalType>()) {
+    TypePtr inner = optional_rhs->getElementType();
+    return can_hold_none_ && types_.size() == 2 &&
+      (types_[0]->isSubtypeOf(inner) ^ types_[1]->isSubtypeOf(inner));
+  } else if (Type::isSubtypeOfExt(rhs_, why_not)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+std::string UnionType::str() const {
+  std::stringstream ss;
+  ss << "Union[";
+  for (size_t i = 0; i < types().size(); ++i) {
+    if (i > 0) {
+      ss << ", ";
+    }
+    ss << types()[i]->str();
+  }
+  ss << "]";
+  return ss.str();
+}
+
+UnionTypePtr UnionType::union_of(std::vector<TypePtr>& rhs_types) const {
+  return union_of({rhs_types});
+}
+
+UnionTypePtr UnionType::union_of(const UnionTypePtr rhs) const {
+  std::vector<TypePtr> new_types = this->types().vec();
+  new_types.insert(new_types.end(), rhs->types().begin(), rhs->types().end());
+  return UnionType::create(std::move(new_types));
+}
+
+UnionTypePtr UnionType::intersection_of(std::vector<TypePtr>& rhs_types) const {
+  return intersection_of({rhs_types});
+}
+
+UnionTypePtr UnionType::intersection_of(const UnionTypePtr rhs) const {
+  std::vector<TypePtr> types;
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{rhs->types().begin(), rhs->types().end()};
+  for (auto lhs_type : this->types()) {
+    // O(1) fast path: If we have the same type in `rhs`
+    if (dict.count(lhs_type)) {
+      types.emplace_back(lhs_type);
+    }
+    // If we can find some common supertype with a type in `rhs`
+    for (const TypePtr rhs_type : rhs->types()) {
+      if (auto unified = unifyTypes(lhs_type, rhs_type)) {
```

Comment:
we shouldn't be using `unifyTypes` here, for example intersection of `None`, `int` is an empty set, but unifyTypes will yield `Optional[int]`. Could we add some tests to this function ?